channel_store.go

```diff
@@ -0,0 +1,125 @@
+package sockchat
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+var ErrChannelDoesNotExist = errors.New("channel does not exist")
+var ErrUserNotInChannel = errors.New("user is not member of this channel")
+var ErrUserAlreadyInChannel = errors.New("user is already member of this channel")
+var ErrEmptyChannelName = errors.New("channel's `name` is missing")
+
+type ChannelStore struct {
+	Channels map[string]*Channel
+	lock     sync.RWMutex
+}
+
+func NewChannelStore() (*ChannelStore, error) {
+	return &ChannelStore{Channels: make(map[string]*Channel)}, nil
+}
+
+func (s *ChannelStore) GetChannel(name string) (*Channel, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	channel := s.Channels[name]
+	if channel == nil {
+		return nil, ErrChannelDoesNotExist
+	}
+	return channel, nil
+}
+
+func (s *ChannelStore) CreateChannel(channelName string) error {
+	if channelName == "" {
+		return ErrEmptyChannelName
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if s.Channels[channelName] != nil {
+		return fmt.Errorf("channel `%s` already exists", channelName)
+	}
+	s.Channels[channelName] = &Channel{members: make(map[SockchatUserHandler]bool)}
+	return nil
+}
+
+func (s *ChannelStore) AddUserToChannel(channelName string, user SockchatUserHandler) error {
+	if channelName == "" {
+		return ErrEmptyChannelName
+	}
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	channel := s.Channels[channelName]
+	if channel == nil {
+		return ErrChannelDoesNotExist
+	}
+	if channel.GetMembers()[user] {
```

Comment:
```suggestion
	if channel.HasMember(user) {
```
User membership is currently checked without a lock, because it is done only after the `members` map is returned. To fix it you can simply move the responsibility of checking membership to the channel itself by creating additional method `Channel.HasMember` which will use a lock inside of a channel.

The same comment applies to all occurrences of `channel.GetMembers()[user] `